archive: refactor race fixing logic to prevent potential deadlock

tiborvass opened the pull request:
Signed-off-by: Tibor Vass <tibor@docker.com>

Follow up to #39860

cc @tonistiigi @cpuguy83 @kolyshkin 

modeled after https://github.com/tonistiigi/copy/blob/go-copy/detect/detect.go#L85

cpuguy83 reviewed (approved):
LGTM

thaJeztah commented:
@tiborvass failures look related;

https://ci.docker.com/public/job/moby/job/PR-39920/3/execution/node/207/log/

```
00:03:03.520  failed to register layer: Error processing tar file(read |0: file already closed): 
```

https://ci.docker.com/public/job/moby/job/PR-39920/3/execution/node/212/log/

```
00:09:48.551  error loading frozen images: Error processing tar file(read |0: file already closed): 
```


https://ci.docker.com/public/job/moby/job/PR-39920/3/execution/node/217/log/

```
00:10:32.212  === Failed
00:10:32.212  === FAIL: pkg/archive TestPigz (0.01s)
00:10:32.212      archive_test.go:1336: Tested whether Pigz is used, as it installed
00:10:32.212      archive_test.go:1339: assertion failed: *os.File (*reflect.rtype) != *io.PipeReader (*reflect.rtype)
00:10:32.212  
00:10:32.212  
```

thaJeztah commented:
ping @tiborvass - what's the status on this one?

thaJeztah commented:
ping @tiborvass

andrewhsu commented:
@tiborvass when you get a chance, rebase this to get the latest tests run on this change

thaJeztah commented:
ping @tiborvass

thaJeztah commented:
ping @tiborvass PTAL

thaJeztah commented:
ping @tiborvass PTAL

tiborvass committed:
archive: add TestCmdStreamDeadlock

The following currently fails

```
GO111MODUEL=off go test -run TestCmdStreamDeadlock ./pkg/archive
```

and is fixed by the next commit.

Signed-off-by: Tibor Vass <teabee89@gmail.com>

tiborvass committed:
archive: refactor race fixing logic to prevent potential deadlock

Signed-off-by: Tibor Vass <teabee89@gmail.com>

tiborvass commented:
@thaJeztah Rebased

thaJeztah commented:
Hm... something's either broken (I've seen this error on another PRs, so might not be related), or there was a glitch that we hit. Let me kick CI again.

```
Error: No such image: emptyfs
Loaded image: emptyfs:latest
Running integration-test (iteration 1)
Running /go/src/github.com/docker/docker/integration/build (amd64.integration.build) flags=-test.v -test.timeout=5m  -test.coverprofile=/go/src/github.com/docker/docker/bundles/test-integration/amd64-integration-build-coverage.out
error loading frozen images: read |0: file already closed

=== Failed
=== FAIL: amd64.integration.build  (0.00s)
error loading frozen images: read |0: file already closed
```
